File: 3509.maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k.py
```python
#
# @lc app=leetcode id=3509 lang=python3
#
# [3509] Maximum Product of Subsequences With an Alternating Sum Equal to K
#

# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int], k: int, limit: int) -> int:
        # The alternating sum is bounded by the sum of absolute values.
        # Since nums[i] >= 0, it's bounded by sum(nums).
        # If k is outside this range, it's impossible.
        if abs(k) > sum(nums):
            return -1
        
        offset = 2000  # Offset to handle negative indices in bitmask
        
        # Filter out zeros for the product calculation
        nums_nz = [x for x in nums if x > 0]
        
        # dp[p][0] stores bitmask of achievable sums for product p, expecting next term to be added (+)
        # dp[p][1] stores bitmask of achievable sums for product p, expecting next term to be subtracted (-)
        # Initialize dp table
        dp = [[0, 0] for _ in range(limit + 1)]
        
        # Base case: Empty subsequence has product 1 (multiplicative identity) and sum 0.
        # It expects the first element (index 0) to be added.
        dp[1][0] = 1 << offset
        
        for x in nums_nz:
            # Iterate backwards to treat each number as used at most once per subsequence
            # We only go down to 1 because p*x must be <= limit
            for p in range(limit // x, 0, -1):
                # If we have valid sums for product p expecting '+', 
                # we can append x. New sum: s + x. New state: expecting '-'.
                if dp[p][0]:
                    dp[p * x][1] |= (dp[p][0] << x)
                
                # If we have valid sums for product p expecting '-',
                # we can append x. New sum: s - x. New state: expecting '+'.
                if dp[p][1]:
                    dp[p * x][0] |= (dp[p][1] >> x)
                    
        # Check for the maximum product
        count_ones = nums.count(1)
        for p in range(limit, 0, -1):
            mask0 = dp[p][0]
            mask1 = dp[p][1]
            
            found = False
            
            # Check if k is achievable ending in a state expecting '-'
            # (meaning the last term was added, i.e., at an even index like 0, 2...)
            if (mask1 >> (k + offset)) & 1:
                found = True
            
            # Check if k is achievable ending in a state expecting '+'
            # (meaning the last term was subtracted, i.e., at an odd index like 1, 3...)
            if not found and ((mask0 >> (k + offset)) & 1):
                # Special check for p=1, k=0.
                # The initialization state (empty seq) is p=1, sum=0, state 0.
                # We must ensure we don't return true for the empty sequence itself.
                # A valid non-empty sequence with prod=1, sum=0 requires at least two 1s (e.g. 1-1).
                if p == 1 and k == 0:
                    if count_ones >= 2:
                        found = True
                else:
                    found = True
            
            if found:
                return p
        
        # If no valid subsequence found with product > 0, check if product 0 is possible.
        # This requires a 0 in nums and a subsequence (containing 0) summing to k.
        if 0 in nums:
            # mnz: Mask No Zero [expect +, expect -]
            # mz: Mask With Zero [expect +, expect -]
            mnz = [1 << offset, 0]
            mz = [0, 0]
            
            for x in nums:
                new_mnz = mnz[:]
                new_mz = mz[:]
                
                if x != 0:
                    # Update No Zero masks
                    new_mnz[1] |= mnz[0] << x
                    new_mnz[0] |= mnz[1] >> x
                    # Update With Zero masks
                    new_mz[1] |= mz[0] << x
                    new_mz[0] |= mz[1] >> x
                else:
                    # x is 0. 
                    # Transition from No Zero -> With Zero
                    # Appending 0 (+0) to state 0 -> state 1
                    new_mz[1] |= mnz[0]
                    # Appending 0 (-0) to state 1 -> state 0
                    new_mz[0] |= mnz[1]
                    
                    # Transition from With Zero -> With Zero
                    new_mz[1] |= mz[0]
                    new_mz[0] |= mz[1]
                
                mnz = new_mnz
                mz = new_mz
            
            # Check if k is in either state for mask with zero
            if ((mz[0] >> (k + offset)) & 1) or ((mz[1] >> (k + offset)) & 1):
                return 0
                
        return -1
# @lc code=end
```

File: 3509.maximum-product-of-subsequences-with-an-alternating-sum-equal-to-k.cpp
#include <bits/stdc++.h>
using namespace std;

#
# @lc app=leetcode id=3509 lang=cpp
#
# [3509] Maximum Product of Subsequences With an Alternating Sum Equal to K
#

# @lc code=start
class Solution {
public:
    int maxProduct(vector<int>& nums, int k, int limit) {
        int n = (int)nums.size();
        int maxSum = 12 * n;
        if (k < -maxSum || k > maxSum) return -1;

        int offset = maxSum;
        int range = 2 * maxSum + 1;
        int W = limit / 64 + 1; // number of uint64_t blocks
        size_t SZ = (size_t)range * W;

        vector<uint64_t> dp[2], nxt[2];
        dp[0].assign(SZ, 0); dp[1].assign(SZ, 0);
        nxt[0].assign(SZ, 0); nxt[1].assign(SZ, 0);

        vector<char> activeFlag[2], nextActiveFlag[2];
        activeFlag[0].assign(range, 0); activeFlag[1].assign(range, 0);
        nextActiveFlag[0].assign(range, 0); nextActiveFlag[1].assign(range, 0);

        vector<int> active[2], nextActive[2];

        auto ptrAt = [&](vector<uint64_t>& arr, int sumIdx) -> uint64_t* {
            return arr.data() + (size_t)sumIdx * W;
        };

        auto activateIfNeeded = [&](int p, int sumIdx) {
            if (!nextActiveFlag[p][sumIdx]) {
                nextActiveFlag[p][sumIdx] = 1;
                nextActive[p].push_back(sumIdx);
            }
        };

        auto setBit = [&](uint64_t* bits, int prod) {
            bits[prod >> 6] |= (1ULL << (prod & 63));
        };

        for (int x : nums) {
            // next = dp (skip current element)
            memcpy(nxt[0].data(), dp[0].data(), SZ * sizeof(uint64_t));
            memcpy(nxt[1].data(), dp[1].data(), SZ * sizeof(uint64_t));
            nextActiveFlag[0] = activeFlag[0];
            nextActiveFlag[1] = activeFlag[1];
            nextActive[0] = active[0];
            nextActive[1] = active[1];

            // Start new subsequence with [x]
            if (x <= limit) {
                int sumIdx = offset + x;
                if (0 <= sumIdx && sumIdx < range) {
                    activateIfNeeded(1, sumIdx);
                    uint64_t* dst = ptrAt(nxt[1], sumIdx);
                    setBit(dst, x);
                }
            }

            // Extend existing subsequences
            for (int p = 0; p <= 1; ++p) {
                for (int sumIdx : active[p]) {
                    uint64_t* src = ptrAt(dp[p], sumIdx);

                    int delta = (p == 0 ? x : -x);
                    int dstSum = sumIdx + delta;
                    if (dstSum < 0 || dstSum >= range) continue;
                    int np = 1 - p;

                    if (x == 0) {
                        // product becomes 0
                        activateIfNeeded(np, dstSum);
                        uint64_t* dst = ptrAt(nxt[np], dstSum);
                        setBit(dst, 0);
                        continue;
                    }

                    uint64_t* dst = ptrAt(nxt[np], dstSum);

                    if (x == 1) {
                        // OR whole bitset (product unchanged)
                        activateIfNeeded(np, dstSum);
                        for (int wi = 0; wi < W; ++wi) dst[wi] |= src[wi];
                        continue;
                    }

                    int maxPr = limit / x;
                    bool activated = nextActiveFlag[np][dstSum];

                    for (int wi = 0; wi < W; ++wi) {
                        uint64_t word = src[wi];
                        while (word) {
                            int b = __builtin_ctzll(word);
                            int pr = (wi << 6) + b;
                            word &= (word - 1);
                            if (pr > limit) continue;

                            int newPr;
                            if (pr == 0) newPr = 0;
                            else {
                                if (pr > maxPr) continue;
                                newPr = pr * x;
                            }
                            if (newPr > limit) continue;

                            if (!activated) {
                                activateIfNeeded(np, dstSum);
                                activated = true;
                            }
                            setBit(dst, newPr);
                        }
                    }
                }
            }

            dp[0].swap(nxt[0]);
            dp[1].swap(nxt[1]);
            activeFlag[0].swap(nextActiveFlag[0]);
            activeFlag[1].swap(nextActiveFlag[1]);
            active[0].swap(nextActive[0]);
            active[1].swap(nextActive[1]);
        }

        int targetIdx = offset + k;
        if (targetIdx < 0 || targetIdx >= range) return -1;

        auto getMaxProduct = [&](vector<uint64_t>& arr) -> int {
            uint64_t* bits = ptrAt(arr, targetIdx);
            int lastBits = (limit & 63);
            uint64_t lastMask = (lastBits == 63) ? ~0ULL : ((1ULL << (lastBits + 1)) - 1ULL);

            for (int wi = W - 1; wi >= 0; --wi) {
                uint64_t word = bits[wi];
                if (wi == W - 1) word &= lastMask;
                if (!word) continue;
                int msb = 63 - __builtin_clzll(word);
                int prod = (wi << 6) + msb;
                if (prod <= limit) return prod;
            }
            return -1;
        };

        int a = getMaxProduct(dp[0]);
        int b = getMaxProduct(dp[1]);
        return max(a, b);
    }
};
# @lc code=end
